configure: cleanup arch specific fast options

yfguo opened the pull request:
## Pull Request Description

1. Separate arch related fast options to `sse2`, `avx` and `avx512f`.
The names represent the corresponding compiler "-m" flags. The `sse2`
is enabled by default because it is widely supported. These options
are consolidated in MPL's configure. Note, these flags will only be
added to the CFLAGS when building MPL, not parent library.

2. The MPL_Memcpy_stream function is uninlined to accommodate for this
change. Both AMD and Intel architecture show no performance
difference uninline it (data included in PR).


## Author Checklist
* [ ] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [ ] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [ ] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

yfguo commented:
test:mpich/ch4/ofi

raffenet reviewed (commented):
If I understand correctly, MPL is supposed to prefix all `AC_DEFINE` macros with `MPL_`. Does this need an complementary patch to update various `#ifdef`s?

yfguo commented:
> If I understand correctly, MPL is supposed to prefix all `AC_DEFINE` macros with `MPL_`. Does this need an complementary patch to update various `#ifdef`s?

Yes. I have updated to the commit to include the `MPL_` prefix at where these checks are used. Also added another fix for incorrect detection code.

yfguo commented:
OK. This need a bit more work to cleanup.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
@raffenet The PR should be good to go now.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
Actually I think it is better to move these check out of MPL. Since we need to process the CFLAGS, it might be cleaner if we just do everything in MPICH.

raffenet commented on src/mpl/include/mpl_trmem.h:
```diff
@@ -316,7 +316,7 @@ static inline void MPL_Memcpy_stream(void *dest, const void *src, size_t n)
 #else
 
-#ifdef HAVE_SSE2
+#ifdef HAVE_MM_STREAM_SI128
```
This will be defined in `mpichconf.h`. Will it get included before this header for this to work? Typically we make sure to include `mpl.h` before anything else.

yfguo replied:
I ended up adding a new m4 to deal with the autoheader for generating the correct config.h.in for MPL. Now the pass-through of the definition should work. And the MPL is revert back to use MPL_ prefixed macros.

yfguo commented:
test:mpich/ch4/ofi

hzhou commented on configure.ac:
```diff
@@ -445,6 +445,7 @@ AC_ARG_ENABLE(fast,
         alwaysinline - Force compiler to always inline performance critical routines
         sse      - Use SSE2-based copy routines
         avx      - Use AVX256-based copy routines
+        avx512   - Use AVX512-based copy routines
```
enable avx or avx512 are too ambiguous -- I know, not your fault ;) The instruction set can be used in multiple features and the `--enable-` options should be per-feature. How about something like `--enable-ch4-shm-nontemporal-copy` (shorter would be nice) then inside configure we check for either avx512 or avx?

yfguo replied:
Yes, there options are not very straight forward. There are two set of functionality that MPICH uses: 1) non-temporal store intrinsics with AVX256 or SSE2 (used in POSIX) and 2) regular mem copy with AVX256 or AVX512 (used in mpl/gpu/mpl_gpu_ze.c).

Thinking about the problem I run into before this PR. I think users would expect `--enable-fast` to be portable, some optimization options may have no effect, code built should behave as if those options were not given. I think the biggest problem is the `avx`, `sse` or `avx512` removes the portability of the MPICH build. It checks if AVX can build and run on the building machine, not the target machine. So if I build on a intel machine, configure will detect AVX512 supported and build with it. If this build is then run on a AMD machine, it will fail because the lack of AVX512 instructions. While is not common that login nodes and compute nodes to have drastically different architecture.

Maybe something like `--enable-sse`, `--enable-avx` and `--enable-avx512` will be a better solution. We can note that these options enable the use of corresponding instructions, which limits the portability of the build. `--enable-sse` should be largely portable among x86, `--enable-avx` will be OK for most Intel CPUs and AMD CPUs starting Zen3, and `--enable-avx512` will be the most limited in portability as only a few Intel CPUs and AMD Zen5 (I think) support it.

More feature specific options like `--enable-ch4-shm-nontemporal-copy` or `--enable-ch4-shm-nt-copy` will be useful for a case where we enable AVX256 but do not want to use non-temporal copy. It will not work the other way, where we only want non-temporal copy feature and no other AVX256 features.

So, I think maybe we should just add `--enable-sse`, `--enable-avx` and `--enable-avx512` to turn on/off the usage of these instruction sets. The specific checkings on individual intrinsic maybe redundant. For example, if a x86 machine has AVX256 support, it should always have support for _mm256_stream_si256. It is probably only useful is some older compiler only implemented partial support of AVX256.

I will redo the PR.

yfguo replied:
Or, we can have `--enable-ch4-shm-nt-copy` and `--enable-mpl-avx-move` to check if the corresponding intrinsics can be built without adding `-mavx` or `-mavx512f` to `CFLAGS`, and shift the responsibility of adding these flags to the user who build MPICH. And we can warn user about missing if those feature are enabled but CFLAGS are missing.

Not the most user-friendly choice IMO.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
Here is a summary of my further investigation.

1. CFLAGS detected in MPL configure is not propagated back up to MPICH CFLAGS. It seems we do not want some flags such as visibility control to leak back to MPICH configure. So, arch specific CFLAGS like `-mavx` all configure need would need to checked at MPICH configure and propagated down. I think it is easier to maintain at way. From the MPICH configure perspective, `--enable-fast=avx` will simply mean "MPICH can use AVX". And it does not worry about who or what will be using it.
2. MPL configure can have Intrinsic (e.g. if _mm512_storeu_si512 is available) checks that relies on those CFLAGS. So if upper level configure does not include them, these checks will failed and disable the corresponding feature. For example, MPICH configure passes CFLAGS=-mavx only, MPL will find that _mm512_* are not available, but _mm256_* and below are available.
3. Intrinsic specific configure options like `--enable-mm512-storeu-si512` seem unnecessary. Because it only controls if our MPL code use the intrinsics explicitly (e.g. https://github.com/pmodels/mpich/blob/00af4b112fe9d0f0b658372159dc29939747781e/src/mpl/src/gpu/mpl_gpu_ze.c#L3162), it does not prevent compiler to generate instructions that match to that intrinsic. That is, we can disable usage of `_mm512_storeu_si512`, but the compiler can still generate `vmovdqu32` for memcpy. Unless AVX512F is globally turned by at MPICH level.
4. Feature specific option like `--enable-nt-copy` might be useful (not added in this PR). POSIX already have an environment variable to control if the inter-NUMA path uses non-temporal copy (`MPIR_CVAR_CH4_SHM_POSIX_TOPO_ENABLE`). I am not sure when would user want to disable nt copy at configure time. Need some insight here.

With this the first commit of this PR, here are some examples on how it behaves under normal circumstance.
A. On a x86 CPU.
  1. Default configuration sets `--enable-fast=O2,sse2`. Everything is built with SSE2 instructions enable, non-temporal copy will be enabled because SSE2 has `_mm_stream_si128`, and ZE specific intrinsic usage will be disable as they need at least AVX.
  2. User set `--enable-fast=O2`, MPICH will build without SSE2 support. So not nt copy, everything else is the same as case A.1.
  3. User set `--enable-fast=O2,avx`, MPICH will build with AVX support. NT copy will be enabled and will be using `_mm256_stream_si256`. ZE specific intrinsic usage will be enabled (uses `_mm256_storeu_si256`). 
  4. User set `--enable-fast=O2,sse2,avx`, this is effectively equal to `--enable-fast=O2,avx` because `CFLAGS=-mavx` is a superset of `CFLAGS=-msse2`.

B. On a ARM CPU.
  1. Default `--enable-fast=O2,sse2` on ARM, MPICH will detect that `-msse2` is supported by the compiler and will not be added to CFLAGS. Intrinsic checks in MPL will also return no and disable corresponding codes.

yfguo commented:
@hzhou I think this PR is ready for another review.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

hzhou commented on configure.ac:
```diff
@@ -953,67 +958,78 @@ if test -z "$enable_fast_no_strict_alignment" ; then
-    if test "$pac_cv_found_avx" = "yes" ; then
+    PAC_PUSH_FLAG([CFLAGS])
+    PAC_APPEND_FLAG([-mavx],[CFLAGS])
+    AC_CACHE_CHECK([whether SSE2 is supported by the CPU], pac_cv_found_avx_runnable,[
```
s/SSE2/AVX/

hzhou reviewed (commented):
@yfguo What if we don't make `MPL_Memcpy_stream` an inline function? Then you won't need leak any of these details out of `mpl`, right?

yfguo commented:
Performance of inlined/uninlined MPIR_Memcpy_stream. Only tested osu_latency for 256B-4KB because smaller message size does not use MPL_Memcpy_stream.

## On AMD Zen3
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#467886" vlink="#96607D">


msg size (B) | inlined | uninlined
-- | -- | --




</body>

</html>

## On Intel Icelake
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#467886" vlink="#96607D">


msg size (B) | inlined | uninlined
-- | -- | --
256 | 0.54 | 0.54
512 | 0.62 | 0.58
1024 | 0.7 | 0.67
2048 | 0.84 | 0.82
4096 | 1.15 | 1.12



</body>

</html>

yfguo commented:
test:mpich/ch4/ofi

hzhou reviewed (approved):
LGTM
